exp/addons/controllers/clusterresourceset_controller.go

```diff
@@ -0,0 +1,357 @@
+	clusterResourceSetBinding, err := GetorCreateClusterResourceSetBinding(ctx, r.Client, cluster, clusterResourceSet)
+	if err != nil {
+		return err
+	}
+
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(clusterResourceSetBinding, r.Client)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		// Always attempt to Patch the ClusterResourceSetBinding object after each reconciliation.
+		if err := patchHelper.Patch(ctx, clusterResourceSetBinding); err != nil {
+			r.Log.Error(err, "failed to patch config")
+		}
+	}()
+
+	// Initializes the maps in ClusterResourceSetBinding object if they are uninitialized.
+	InitClusterResourceSetBinding(clusterResourceSetBinding, clusterResourceSet)
+
+	var clusterResourceSetInMap addonsv1.ResourcesSetBinding
+	var ok bool
+	if clusterResourceSetInMap, ok = clusterResourceSetBinding.Spec.Bindings[clusterResourceSet.Name]; !ok {
+		// This should never be the case as InitClusterResourceSetBinding is adding the ClusterResourceSet to the map during initialization.
+		return errors.New("failed to find ClusterResourceSet " + clusterResourceSet.Name + " in ClusterResourceBinding")
+	}
+
+	errList := []error{}
+
+	// Iterate all resources and apply them to the cluster and update the resource status in the ClusterResourceSetBinding object.
+	for _, resource := range clusterResourceSet.Spec.Resources {
+		// If resource is already applied successfully and clusterResourceSet mod is "ApplyOnce", continue. (No need to check hash changes here)
+		resourceInMap := addonsv1.ResourceBinding{}
+		if resourceInMap, ok = clusterResourceSetInMap.Resources[GenerateResourceKindName(resource)]; ok {
+			if clusterResourceSet.Spec.Strategy == string(addonsv1.ClusterResourceSetStrategyApplyOnce) && resourceInMap.Applied {
+				continue
+			}
+		}
+
+		// This is set here to cover unsuccessful apply scenarios (if the end of the loop is not reached)
+		resourceInMap = *GenerateResourceStatus(false, "")
+		clusterResourceSetInMap.Resources[GenerateResourceKindName(resource)] = resourceInMap
+
+		// Only allow using resources (Secrets/Configmaps) in the same namespace with the cluster.
+		typedName := types.NamespacedName{Name: resource.Name, Namespace: cluster.Namespace}
+		dataList := make([][]byte, 0)
+
+		// Retrieve data in the resource as an array because there can be many key-value pairs in resource and all will be applied to the cluster.
+		switch resource.Kind {
+		case string(addonsv1.ConfigMapClusterResourceSetResourceKind):
+			// Set status in ClusterResourceSetBinding in case of early continue due to a failure. Do this for only supported kinds.
+			clusterResourceSetInMap.Resources[GenerateResourceKindName(resource)] = addonsv1.ResourceBinding{
+				Hash:            "",
+				Applied:         false,
+				LastAppliedTime: &metav1.Time{Time: time.Now().UTC()},
+			}
+
+			resourceConfigMap, err := GetConfigMapFromNamespacedName(context.Background(), r.Client, typedName)
```

Comment:
If I got this right, there is no ownerRef between the cluster resource set and the ConfigMap (or the secret); if this is true, this means that clusterctl move in its current form won't be capable to detect which ConfigMap (or the secret) should be moved together with a cluster.

Reply:
> Does it make sense for the CRS to own the ConfigMaps and Secrets?

The need for an owner ref chain was raised also on the proposal, because the move algorithms rely on owner references to determine what is in scope and the order of the create/delete operations.

Accordingly, https://github.com/kubernetes-sigs/cluster-api/pull/3243 is implemented on top of this assumption
- ClusterResourceSet owns Secrets and ConfigMaps

@sedefsavas. @ncdc is this approach acceptable? 

Please note that at discovery time clusterctl only stores a small set of info about each object (GVK, namespace, name, uid, and ownerReferences), so, as of today, the object spec/the resource list in ClusterResourceSet can't be used to identify links between objects; 

I would advise against implementing a custom approach for CRS, due to the complexity of move and to the fact that this might be an invasive change, but I'm waiting for feedbacks about how to proceed

Reply:
It makes sense to add CRS as owner since we use a CRS specific kind when we create secret resources for example. We can assume resource ConfigMaps and Secrets are only for CRS consumption. This will be beneficial especially when we implement `Sync` mode, we may want to detect changes on the resources and using ownerRef, we can map them to CRSs. 
I am adding CRS as ownerRef for now, will revert the changes if there are concerns. 

@fabriziopandini is it okay to have multiple CRS owner refs per Secret/ConfigMap for clusterctl? Resources may be shared among CRSs.

Reply:
>  is it okay to have multiple CRS owner refs per Secret/ConfigMap for clusterctl

yes, I don't see problems in having multiple owner refs

BTW, I managed to prototype a solution for handling the lack of owner ref on Secrets/ConfigMaps, not ideal, but hopefully acceptable, so feel free to choose the option that best suits this problem

Reply:
Sounds good, @ncdc considering my points above, does it make sense to assign CRS as resource ownerref?

Reply:
Added CRS to resources as owner ref.

Reply:
To follow up ownership discussions: https://github.com/kubernetes-sigs/cluster-api/issues/3301